src/trails.erl

```diff
@@ -1,35 +1,78 @@
 -module(trails).
 
-
 -export([single_host_compile/1]).
 -export([compile/1]).
 -export([trail/2]).
 -export([trail/3]).
 -export([trail/4]).
+-export([trail/5]).
+
+-opaque trail() ::
+  #{ path_match  => integer()
+   , constraints => integer()
+   , handler     => module()
+   , options     => any()
+   , metadata    => map()
+   }.
+-export_type([trail/0]).
+
+-type trails() :: [ trails:trail() | cowboy_router:route_path() ].
+-export_type([trails/0]).
 
 -spec single_host_compile([cowboy_router:route_path()]) ->
   cowboy_router:dispatch_rules().
 single_host_compile(Trails) ->
   compile([{'_', Trails}]).
 
--spec compile(cowboy_router:routes()) -> cowboy_router:dispatch_rules().
+-spec compile(trails()) -> cowboy_router:dispatch_rules().
```

Comment:
I think `trails()` is not the right type here, since it's missing the host part… maybe it would be better to change it to `[{Host::cowboy_router:route_match(), Trails::trails()}]`